Add --readonly-app-data option

asottile opened the pull request:
Resolves #2000

### Thanks for contributing, make sure you address all the checklists (for details on how see
[development documentation](https://virtualenv.pypa.io/en/latest/development.html#development))!

- [x] ran the linter to address style issues (``tox -e fix_lint``)
- [x] wrote descriptive pull request text
- [x] ensured there are test(s) validating the fix
- [x] added news fragment in ``docs/changelog`` folder
- [x] updated/extended the documentation

asottile commented:
hmm this still doesn't quite work -- need to adjust this bit here to use the app_data's "locking" mechanism https://github.com/pypa/virtualenv/blob/9708b54f0f1fe4aad145c12842fae241707d696c/src/virtualenv/seed/embed/via_app_data/via_app_data.py#L55

asottile commented:
hmmm maybe that last bit isn't correct, I'll take a look at this tomorrow

asottile commented:
I believe this is ready for review now 🎉

gaborbernat commented on src/virtualenv/run/__init__.py:
```diff
@@ -88,32 +89,32 @@ def handle_extra_commands(options):
+    parser.add_argument(
+        "--readonly-app-data",
+        action="store_true",
+        help="allow reading from the configured app data even if the app data directory is readonly",
```
We're not actually checking the app data is read-only, and we really don't care if it's or not. We just enforce it via logic that we don't update it, so a better message here is `use app data folder in read-only mode (write operations will fail with error)`

gaborbernat commented on src/virtualenv/run/__init__.py:
```diff
@@ -88,32 +89,32 @@ def handle_extra_commands(options):
     options, _ = parser.parse_known_args(args, namespace=options)
-    if options.app_data == "<temp folder>":
-        options.app_data = TempAppData()
+
```
This empty line should have been removed.

gaborbernat reviewed (changes requested):
The test suite on Windows 2.7 seems to be failing. The delta coverage is `90%`, can you please bring it up to `100%`? Otherwise conceptually the PR looks good.

asottile commented on src/virtualenv/app_data/__init__.py:
```diff
@@ -5,58 +5,53 @@
+
+def make_app_data(folder, **kwargs):
+    readonly = kwargs.pop("readonly")
+    if kwargs:  # TODO: py3+ kwonly
```
yes, when you're python3 only this can be a keyword-only argument

gaborbernat replied:
Can we even handle that in a 2/3 code?

asottile replied:
what I've implemented is 2+3-compatible keyword-only argument -- when virtualenv eventually drops python2 this can use the actual syntax

gaborbernat replied:
> when virtualenv eventually drops python2

I see. Drop the TODO marker then please, I don't like using them. We don't plan to be python 3 only for the foreseeable future (at least until pypy2 becomes EOL).

asottile commented on src/virtualenv/app_data/__init__.py:
```diff
@@ -5,58 +5,53 @@
-            if folder is not None:
-                return AppDataDiskFolder(folder)
-        return AppDataDisabled()
+    if readonly:
```
hmm I was going based on the c# keyword or html attribute, but I can sed

gaborbernat replied:
It's clearer if you invoke it via a flag, but confusing for config/env-var users, so I'd rather not go down that path.

asottile commented on src/virtualenv/app_data/readonly.py:
```diff
@@ -0,0 +1,31 @@
+
+class PyInfoStoreDiskReadonly(PyInfoStoreDisk):
+    def write(self, content):
+        raise RuntimeError("--readonly-app-data: attempted to write py_info")
```
hmmm so I chose to put the option name here so it's more clear why the error is occuring, but if you want I can change these

asottile commented:
here's the diff cover:

- src/virtualenv/app_data/__init__.py (84.0%): Missing lines 22,28,41-42
    - (previously uncovered, just moved) https://github.com/pypa/virtualenv/pull/2009/files#diff-e2c6ee2a445e2be45e7e528ab167451b93164211148cc9a3ce05a57e8e79ddd8R22
    - (unreachable / defensive): https://github.com/pypa/virtualenv/pull/2009/files#diff-e2c6ee2a445e2be45e7e528ab167451b93164211148cc9a3ce05a57e8e79ddd8R28
    - (previously uncovered, just moved) https://github.com/pypa/virtualenv/pull/2009/files#diff-e2c6ee2a445e2be45e7e528ab167451b93164211148cc9a3ce05a57e8e79ddd8R41-R42
- src/virtualenv/app_data/base.py (100%)
- src/virtualenv/app_data/na.py (100%)
- src/virtualenv/app_data/read_only.py (84.2%): Missing lines 13,17,31
    - (trivial error case): https://github.com/pypa/virtualenv/pull/2009/files#diff-e2c6ee2a445e2be45e7e528ab167451b93164211148cc9a3ce05a57e8e79ddd8R41-R42
    - (trivial error case): https://github.com/pypa/virtualenv/pull/2009/files#diff-e2c6ee2a445e2be45e7e528ab167451b93164211148cc9a3ce05a57e8e79ddd8R41-R42
    - (trivial error case): https://github.com/pypa/virtualenv/pull/2009/files#diff-e2c6ee2a445e2be45e7e528ab167451b93164211148cc9a3ce05a57e8e79ddd8R41-R42
- src/virtualenv/app_data/via_disk_folder.py (80.0%): Missing lines 56
    - (`__repr__` implementation to make debugging easier): https://github.com/pypa/virtualenv/pull/2009/files#diff-e2c6ee2a445e2be45e7e528ab167451b93164211148cc9a3ce05a57e8e79ddd8R41-R42
- src/virtualenv/app_data/via_tempdir.py (100%)
- src/virtualenv/run/__init__.py (100%)
- src/virtualenv/seed/embed/via_app_data/via_app_data.py (100%)
- src/virtualenv/seed/wheels/bundle.py (100%)
- src/virtualenv/util/lock.py (100%)

gaborbernat commented on src/virtualenv/util/lock.py:
```diff
@@ -113,8 +139,30 @@ def lock_for_key(self, name, no_block=False):
 __all__ = (
-    "Timeout",
-    "ReentrantFileLock",
+    NoOpFileLock.__name__,
```
Eh, please no.

gaborbernat commented on src/virtualenv/util/lock.py:
```diff
@@ -113,8 +139,30 @@ def lock_for_key(self, name, no_block=False):
+            yield
+
+
+class NoOpFileLock(FileLockBase):
```
Calling this file lock feels a bit cheating, maybe we want to move the interface to a simple Lock, and then inherit this from that?

asottile replied:
isn't that what this does?

gaborbernat replied:
What I'm saying is:

```python
class Lock(ABC):
     pass

class FileLock(Lock):
     pass

class NoOpLock(Lock):
   pass
```

asottile replied:
I don't see how renaming `FileLockBase` improves anything but I'll do it if you want

gaborbernat replied:
Yeah, let's do that please.

gaborbernat commented on tests/unit/seed/embed/test_bootstrap_link_via_app_data.py:
```diff
@@ -125,6 +143,55 @@ def test_base_bootstrap_link_via_app_data_not_writable(tmp_path, current_fastest
     assert result
 
 
+@pytest.mark.skipif(sys.platform == "win32", reason="Windows only applies R/O to files")
```
Seems like we still should run the test, but mark read only the files themselves, and not the dir, not?

asottile replied:
marking files read-only (and not directories) only prevents appending not deleting / creating / replacing so it's not a useful test

the reason I had to make *both* directories and files read-only is making a root directory read-only doesn't prevent writes in subdirectories

gaborbernat replied:
So you're saying that pre-commit CI will not run on Windows then?

asottile replied:
correct it won't

but it doesn't matter -- you could reproduce this same test on windows but it involves creating an admin user and a non-admin user and I don't think you want to do that in your testsuite :)

asottile replied:
additionally, without my updates the test right above this wasn't actually testing anything 😆

gaborbernat replied:
Cool.

gaborbernat commented on src/virtualenv/run/__init__.py:
```diff
@@ -1,8 +1,9 @@
 from __future__ import absolute_import, unicode_literals
 
+import functools
```
Ping on this.

asottile replied:
oh I thought the shrug was "I don't care either way"

gaborbernat reviewed (approved):
![](https://media.tenor.co/images/15c955585420695acab185a96d094a82/tenor.gif)
